File: c/binary_search.rb
# binary_search.rb

# Load the embedded C extension
require_relative 'binary_search_ext'

# Define a Ruby method that delegates to the embedded C binary search function
def binary_search(arr, target)
  binary_search_c(arr, target)
end

# Usage example
sorted_array = [1, 3, 12, 15, 22, 27, 32]
target = 22
index = binary_search(sorted_array, target)
if index != -1
  puts "The target #{target} is found at index #{index}."
else
  puts "The target #{target} is not found in the array."
end

File: _js/binarysearch.js
function binarySearch(arr, target) {
  let low = 0, high = arr.length - 1;

  while (low <= high) {
    const mid = low + Math.floor((high - low) / 2);
    if (arr[mid] === target) return mid;
    if (arr[mid] < target) low = mid + 1;
    else high = mid - 1;
  }

  return -1;
}

// Example usage:
const arr = [1, 7, 8, 12, 14, 22, 28, 35, 41];
const target = 12;
console.log(binarySearch(arr, target)); // Output: 3 (index of the target element)

function bsearch(arr,target){
  let low=0;
  let high=arr.length-1;

  while (low <= high) {
    const mid=low+Math.floor((high-low)/2);
    if (arr[mid]===target) return mid;
    if (arr[mid]<target) low=mid+1;
    else high=mid-1;
  }
  return -1
}
console.log(binarySearch(arr, target)); // Output: 3 (index of the target element)
